File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    //do a post-order traversal because we need to process child first than the node.
    //if we get positive subtree sum we can include to that current node sum else ignore it.
    int max_sum=INT_MIN;
    
    int gainsum(TreeNode* root){
        if(root==NULL){
            return 0;
        }
        int gain_left=max(gainsum(root->left),0);
        int gain_right=max(gainsum(root->right),0);
        max_sum=max(max_sum,gain_left+gain_right+root->val);
        return max(gain_left+root->val,gain_right+root->val);
    }
    int maxPathSum(TreeNode* root) {
        gainsum(root);
        return max_sum;
    }
};

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        ans=float("-inf")
        def getmax(root):
            nonlocal ans
            if root is None:
                return 0
            gain_left=max(getmax(root.left),0)
            gain_right=max(getmax(root.right),0)
            temp=root.val+gain_left+gain_right
            ans=max(temp,ans)
            return max(gain_left,gain_right)+root.val
        getmax(root)
        
        return ans
        
        